### PR TITLE
Reland of a bunch of pattern matcher + indexing improvements

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112476

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

re: the optimization landed in this PR, see https://github.com/pytorch/pytorch/pull/112102#issuecomment-1780265536

### CODE REVIEW

**Comment by Chillee** [comment]
@chillee has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50826866).

File: torch/_inductor/fx_passes/post_grad.py [hunk 1]
```diff
@@ -657,7 +666,7 @@ def can_replace(node, mutated_arg):
             if len(shared_view_nodes) > 2:  # Arg aliases another node other than copy_
                 return False
 
-            # Check for any uses other than current node and copy_ epilogue
+            # # Check for any uses other than current node and copy_ epilogue
```
Comment (oulgen) [inline torch/_inductor/fx_passes/post_grad.py:669]: you commented your comment

File: torch/_inductor/lowering.py [hunk 2]
```diff
@@ -2845,6 +2846,7 @@ def index_impl(x, indices, check):
         indices_loaders,
         indexed_size,
         x_loader,
+        check=check,
```
Comment (oulgen) [inline torch/_inductor/lowering.py:2849]: did we not rename this?

**Review by oulgen — APPROVED** [review]

**Pull request closed without merging**